test/cleanup.go

```diff
@@ -22,16 +22,36 @@ package test
 import (
 	"os"
 	"os/signal"
+	"sync"
 )
 
+// cleanFuncs stores all the functions to cleanup
+type cleanFuncs struct {
+	once sync.Once
+	s    *sync.Map
+}
+
+func init() {
+	cf = &cleanFuncs{s: &sync.Map{}, once: sync.Once{}}
+}
+
+var cf *cleanFuncs
+
 // CleanupOnInterrupt will execute the function cleanup if an interrupt signal is caught
 func CleanupOnInterrupt(cleanup func()) {
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, os.Interrupt)
+	cf.s.Store(&cleanup, struct{}{})
 	go func() {
 		for range c {
-			cleanup()
-			os.Exit(1)
+			cf.once.Do(func() {
+				cf.s.Range(func(clean interface{}, _ interface{}) bool {
```

Comment:
We should process cleanups in the reverse order they were queued, but I'd wager this is random given Go's (normal) map iteration semantics.

Reply:
yes, this is most likely random

Reply:
@mattmoor @vagababov Did you mean that Go's map iteration is random but we should process cleanups in the reverse order anyway? Or Go's map iteration is random so we compromise the cleanups in the reverse order?

Reply:
Yes map iteration is random, thus the loop is random